lib/fallback.ts

```diff
@@ -10,7 +10,72 @@ export type FallbackOptions = {
  * Allows to provide a substitute value when method call fails or rejects.
  * @param options optional fallback options
  */
-export function fallback(valueProvider: (...args: any[]) => any, options?: FallbackOptions);
-export function fallback(value: any, options?: FallbackOptions) {
-  throw new Error('Not implemented.');
+export function fallback(valueProvider: (...args: any[]) => any, options?: FallbackOptions): any;
+export function fallback(value: any, options?: FallbackOptions): any;
+export function fallback(value: any | ((...args: any[]) => any), options?: FallbackOptions): any {
+
+  return function (target: any, propertyKey: any, descriptor: PropertyDescriptor) {
+    const method = descriptor.value;
+
+    descriptor.value = function () {
+      try {
+        const result = method.call(this, arguments);
+        const isPromiseLike = result && typeof result.then === 'function';
+
+        if (isPromiseLike) {
+          return fallbackPromise(options, value, result, this);
+        }
+
+        return result;
+
+      } catch (err) {
+        const isFiltered = filterError(err, options, this);
+
+        if (isFiltered) {
+          return getFallbackValue(value, this);
+        }
+
+        throw err;
+      }
+    };
+
+  };
+}
+
+function fallbackPromise(
+  options: FallbackOptions,
+  value: any,
+  result: any,
+  instance: any,
+): Promise<any> {
+
+  const resolve = (response: any) => Promise.resolve(response);
+
+  const reject = (err: any) => {
+    const isFiltered = filterError(err, options, instance);
+
+    return isFiltered
+      ? Promise.resolve(getFallbackValue(value, instance))
+      : Promise.reject(err);
+  };
+
+  return result.then(resolve, reject);
```

Comment:
You can avoid use of `resolve`, delete it completely and instead use `catch`: `return result.catch(reject);`